collector/set.go

```diff
@@ -0,0 +1,110 @@
+package collector
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"sync/atomic"
+
+	"github.com/giantswarm/microerror"
+	"github.com/giantswarm/micrologger"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+type SetConfig struct {
+	Collectors []Interface
+	Logger     micrologger.Logger
+}
+
+type Set struct {
+	collectors []Interface
+	logger     micrologger.Logger
+
+	bootedCounter uint32
+	mutex         sync.Mutex
+}
+
+func NewSet(config SetConfig) (*Set, error) {
+	if len(config.Collectors) == 0 {
+		return nil, microerror.Maskf(invalidConfigError, "%T.Collectors must not be empty", config)
+	}
+	if config.Logger == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.Logger must not be empty", config)
+	}
+
+	s := &Set{
+		collectors: config.Collectors,
+		logger:     config.Logger,
+
+		bootedCounter: 0,
+		mutex:         sync.Mutex{},
+	}
+
+	return s, nil
+}
+
+func (s *Set) Boot(ctx context.Context) error {
+	s.logger.Log("level", "debug", "message", "booting collector")
+
+	if s.isBooted() {
+		return nil
+	}
```

Comment:
Why not boot.Once?

Reply:
I wrote that in another comment. I think it was in the first proposal in `operatorkit`. I want the `Boot` interface to return an error. This is not really nicely possible with `sync.Once`.